src/id.rs

```diff
@@ -0,0 +1,96 @@
+use crate::ed25519;
+use bs58;
+use disco::DiscoHash;
+use std::hash::Hash;
+use std::{convert::TryFrom, fmt, str::FromStr};
+use uint::*; // U256
+
+/// NodeId Struct
+///
+/// - wrapper around DiscoHash
+/// - default implements the trait described below
+/// - easy TODO: refactor to make more generic (don't specify DiscoHash)
+pub struct NodeId<T: Hash + Eq + PartialEq + Clone> {
+    pub id: T,
```

Comment:
add preimage and post-image like key, but decouple from ID. Many types of keys and many types of IDs. Two way surjective mappings